src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.WorkerThread.cs

```diff
@@ -0,0 +1,164 @@
+                        if (oldCounts == counts)
+                        {
+                            HillClimbing.ThreadPoolHillClimber.ForceChange(newCounts.numThreadsGoal, HillClimbing.StateOrTransition.ThreadTimedOut);
+                            // TODO: Event:  Worker Thread stop event
+                            return;
+                        }
+                        counts = oldCounts;
+                    } 
+                }
+            }
+
+            internal static void MaybeAddWorkingWorker()
+            {
+                ThreadCounts counts = ThreadCounts.VolatileReadCounts(ref s_aligned.counts);
+                ThreadCounts newCounts;
+                while (true)
+                {
+                    newCounts = counts;
+                    newCounts.numProcessingWork = Math.Max(counts.numProcessingWork, Math.Min((short)(counts.numProcessingWork + 1), counts.numThreadsGoal));
+                    newCounts.numExistingThreads = Math.Max(counts.numExistingThreads, newCounts.numProcessingWork);
+                    
+                    if(newCounts == counts)
+                    {
+                        return;
+                    }
+
+                    ThreadCounts oldCounts = ThreadCounts.CompareExchangeCounts(ref s_aligned.counts, newCounts, counts);
+
+                    if(oldCounts == counts)
+                    {
+                        break;
+                    }
+
+                    counts = oldCounts;
+                }
+
+                int toCreate = newCounts.numExistingThreads - counts.numExistingThreads;
+                int toRelease = newCounts.numProcessingWork - counts.numProcessingWork;
+
+                if(toRelease > 0)
+                {
+                    s_semaphore.Release(toRelease);
+                }
+
+                for (int i = 0; i < toCreate; i++)
+                {
+                    CreateWorkerThread();
+                }
+            }
+
+            internal static bool ShouldStopProcessingWorkNow()
+            {
+                ThreadCounts counts = ThreadCounts.VolatileReadCounts(ref s_aligned.counts);
+                while (true)
+                {
+                    if (counts.numExistingThreads <= counts.numProcessingWork)
+                    {
+                        return true;
+                    }
+
+                    ThreadCounts newCounts = counts;
+                    newCounts.numProcessingWork--;
+
+                    ThreadCounts oldCounts = ThreadCounts.CompareExchangeCounts(ref s_aligned.counts, newCounts, counts);
+
+                    if (oldCounts == counts)
+                    {
+                        return false;
+                    }
+                    counts = oldCounts;
+                }
+            }
+
+            private static void CreateWorkerThread()
+            {
+                RuntimeThread workerThread = RuntimeThread.Create(WorkerThreadStart);
```

Comment:
One disadvantage (from performance point of view) of using RuntimeThread.Create/Start here is that the parent thread will wait for the child thread to start. In addition, I can see there is code in your change that calls CreateWorkerThread in a loop. So it will literary start one thread at a time.